### PR TITLE
gitserver: allow missing date in for-each-refs tags output

### LINKED ISSUE #33446: gitserver: Empty creatordate leads to bad date format error
We're seeing an error in prod for torvalds/Linux: ([logs](https://sourcegraph.grafana.net/explore?orgId=1&left=%7B%22datasource%22:%22grafanacloud-sourcegraph-logs%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Bapp%3D%5C%22worker%5C%22,namespace%3D%5C%22prod%5C%22%7D%20%7C%3D%20%5C%22commit%5C%22%22%7D%5D,%22range%22:%7B%22from%22:%22now-5m%22,%22to%22:%22now%22%7D%7D))

```
gitserver.RefDescriptions: unexpected output from git for-each-ref (bad date format) \"c39ae07f393806ccf406ef966e9a15afc43cc36arefs/tags/v2.6.11 \"
```

It seems like the problem is that somehow, one is ending up an empty value for `%(creatordate)`, which the code doesn't handle. It's not super clear when why that is happening here; tags seem to act as roots for git gc, so resetting HEAD, rebasing etc. don't end up affecting the `creatordate`. Maybe there's some other git operation that leaves tags behind pointing to a non-existent commit?

One unusual thing is that `c39ae07f393806ccf406ef966e9a15afc43cc36a` isn't a valid commit hash for torvalds/linux right now. However, the GitHub UI does show `v2.6.11` as a valid tag. However, if you click on that tag, it gives a 404.

Linux repo showing v2.6.11 tag
Linux repo showing 404 on navigating to v2.6.11 tag

### PR DESCRIPTION
**Pull request by Strum355**
Closes #33446 

## Test plan

Added test-case to `TestParseRefDescriptions`.

### CODE REVIEW

File: internal/vcs/git/commits.go [hunk 1]
```diff
@@ -787,10 +787,18 @@ lineLoop:
...
+			createdDatePtr  *time.Time
+		)
+		createdDate, err := time.Parse(time.RFC3339, createdDatePart)
+		if err != nil && createdDatePart != "" {
 			return nil, errors.Errorf(`unexpected output from git for-each-ref (bad date format) "%s"`, line)
 		}
+		createdDatePtr = &createdDate
+		if createdDate.Equal(time.Time{}) {
+			createdDatePtr = nil
+		}
```
Comment (efritz) [inline internal/vcs/git/commits.go:801]:
```suggestion
		if createdDatePart != "" {
			createdDate, err := time.Parse(time.RFC3339, createdDatePart)
			if err != nil {
				return nil, errors.Errorf(`unexpected output from git for-each-ref (bad date format) "%s"`, line)
			}
			createdDatePtr = &createdDate
		}
```

File: internal/gitserver/gitdomain/common.go [hunk 2]
```diff
@@ -101,5 +101,5 @@ type RefDescription struct {
 	Name            string
 	Type            RefType
 	IsDefaultBranch bool
-	CreatedDate     time.Time
+	CreatedDate     *time.Time
```
[the remark this answers is not in the document]
Comment (Strum355) [reply internal/gitserver/gitdomain/common.go:104]: We _do_ ultimately skip these on the consumer side, Im not sure skipping them on the producer side is the correct side to make that decision on

[the remark this answers is not in the document]
Comment (efritz) [reply internal/gitserver/gitdomain/common.go:104]: Oh is that true? Would the better solution to be not make this a pointer and just skip the commits with no created date?

[the remark this answers is not in the document]
Comment (efritz) [reply internal/gitserver/gitdomain/common.go:104]: Possibly simpler solution: put a `continue` in the else of https://github.com/sourcegraph/sourcegraph/pull/33550#discussion_r844105886

[the remark this answers is not in the document]
Comment (Strum355) [reply internal/gitserver/gitdomain/common.go:104]: Im fine with either solution ultimately, I just feels more correct for gitserver API to not make that decision for us (a source of much frustration for me when APIs try to second guess you)

[the remark this answers is not in the document]
Comment (efritz) [reply internal/gitserver/gitdomain/common.go:104]: I'm also ok with this implementation for an immediate fix and the change has a small blast radius. I also did specifically ask for @Strum355 to have a solution that didn't impose what nil meant.

**Review by efritz — APPROVED** [review]

**Comment by efritz** [comment]
I'd say this is good to merge for now but we can follow-up with moving the decision to skip the ref description object closer to the gitserver request.

**Comment by varungandhi-src** [comment]
> I'd say this is good to merge for now but we can follow-up with moving the decision to skip the ref description object closer to the gitserver request.

Is it pressing to merge this for fixing the errors we are seeing? (Maybe I'm misunderstanding the urgency.)

---

The details of the patch aside, I feel like there is a bunch of valuable context that this patch should be including, but it isn't. For example, consider the next person who looks at the code and what they need to do to understand why we are handling the empty string case/when that can happen:
1. They need to look at git blame for when the change was added.
2. They may come to this PR and not find anything.
3. They end up looking at the issue and see your comment. This provides a 90% explanation but not 100%.

That is 3 hops, and still the information isn't "distilled" into what the core problem is which is roughly something like:

> Some repositories attach tags to non-commit objects, such as trees. This is supposed to be ill-formed according to the Git documentation on tags (<add link>) but this doesn't seem to be validated. In such a situation, one cannot "deference" the tag to obtain the commit it points to.

**Comment by efritz** [comment]
> Is it pressing to merge this for fixing the errors we are seeing?

As far as I understand this context, this is the root cause of the alert we silenced yesterday is because of this repo continuously running into an empty date string.

Seems like there are two solutions:

(a) Filter out in commits with empty dates in the `RefDescriptions` method in the internal/vcs/git package (assumes that's what the user wants)
(b) Make empty date a legal output of for-each-ref (which ... it clearly is) so that the user can determine what to do with it. In this case, we also skip it.

This PR does **(b)**, but we could easily change it to **(a)** pre or post-merge without affecting much.

> I feel like there is a bunch of valuable context that this patch should be including, but it isn't.

Thanks for flagging this, @varungandhi-src. It would be super helpful to put this directly where we check for an empty creatordate. Otherwise it seems like we're overly-lenient as a possible mistake, and that could lead us back to a refactor of the previous behavior.

That comment Varun left is gold - please include it :simple_smile:.

**Comment by varungandhi-src** [comment]
Sure, I hadn't realized that un-silencing the alert was high priority, but it makes sense. 👍 but please follow-up on this; I don't think we should push the complexity of handling the `nil` case to consumers unless there is a compelling reason to do so.

**Review by varungandhi-src — APPROVED** [review]

**Comment by Strum355** [comment]
Have pushed an inline comment. Of note, the docs actually _do_ state that tagging non-commit objects is allowed:
> git tag [-a | -s | -u \<keyid\>] [-f] [-m \<msg\> | -F \<file\>] [-e] \<tagname\> ***[\<commit\> | \<object\>]***
> ...
> The object that the new tag will refer to, usually a commit.

**Comment by varungandhi-src** [comment]
Apparently, the build was still going for after an hour. https://buildkite.com/sourcegraph/sourcegraph/builds/140927

I stopped and restarted it.

**Comment by Strum355** [comment]
Thats annoying, I had already had to restart coz it was stuck. Wondering if i need to merge from main to catch some changes..

File: internal/vcs/git/commits.go [hunk 3]
```diff
@@ -782,9 +782,18 @@ lineLoop:
...
 
-		createdDate, err := time.Parse(time.RFC3339, string(parts[3]))
-		if err != nil {
-			return nil, errors.Errorf(`unexpected output from git for-each-ref (bad date format) "%s"`, line)
+		var (
+			createdDatePart = string(parts[3])
+			createdDatePtr  *time.Time
+		)
+		// Some repositories attach tags to non-commit objects, such as trees. In such a situation, one
+		// cannot deference the tag to obtain the commit it points to, and there is not associated creatordate.
```
Comment (ryanslade) [inline internal/vcs/git/commits.go:790]:
```suggestion
		// cannot deference the tag to obtain the commit it points to, and there is no associated creatordate.
```

**Review by ryanslade — APPROVED** [review]
Thanks for fixing this!

**Pull request merged**